core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java

```
@@ -0,0 +1,300 @@
+    int bytesRead = read(mSingleByte);
+    if (bytesRead == -1) {
+      return -1;
+    }
+    Preconditions.checkState(bytesRead == 1);
+    return BufferUtils.byteToInt(mSingleByte[0]);
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkArgument(len >= 0, "length should be non-negative");
+    Preconditions.checkArgument(off >= 0, "offset should be non-negative");
+    if (len == 0) {
+      return 0;
+    }
+    if (mPosition >= mStatus.getLength()) { // at end of file
+      return -1;
+    }
+    int bytesRead = 0;
+    // for each page, check if it is available in the cache
+    while (bytesRead < len && mPosition < mStatus.getLength()) {
+      long currentPage = mPosition / mPageSize;
+      int currentPageOffset = (int) (mPosition % mPageSize);
+      int bytesLeftInPage = (int) Math.min(mPageSize - currentPageOffset, len - bytesRead);
+      PageId pageId = new PageId(mStatus.getFileId(), currentPage);
+      try (ReadableByteChannel cachedData = mCacheManager.get(pageId, currentPageOffset)) {
+        if (cachedData != null) { // cache hit
+          // wrap return byte array in a bytebuffer and set the pos/limit for the page read
+          ByteBuffer buf = ByteBuffer.wrap(b);
+          buf.position(off + bytesRead);
+          buf.limit(off + bytesRead + bytesLeftInPage);
+          // read data from cache
+          while (buf.position() != buf.limit()) {
+            if (cachedData.read(buf) == -1) {
+              break;
+            }
+          }
+          Preconditions.checkState(buf.position() == buf.limit());
+          bytesRead += bytesLeftInPage;
+          mPosition += bytesLeftInPage;
+        } else { // cache miss
+          byte[] page = readExternalPage(mPosition);
+          if (page.length > 0) {
+            mCacheManager.put(pageId, page);
```

how shall we handle failed put with IOE? shall we silently swallow the exception (w/ warning)?